File: leetcode/求和/python/combinationSum.py
class Solution:

    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers

    def combinationSum(self, candidates, target):
        candidates = sorted(list(set(candidates)))

        def binarysearch(end, num):
            start = 0
            while start < end:
                mid = (start + end + 1) >> 1
                if candidates[mid] > num:
                    end = mid - 1
                else:
                    start = mid
            return start

        result, i = [], binarysearch(len(candidates) - 1, target)
        stack, residue = [i], target - candidates[i]
        while stack:
            if residue > 0:
                j = stack[-1]
                if residue >= candidates[j]:
                    n, residue = divmod(residue, candidates[j])
                    stack += [j] * n
                else:
                    k = binarysearch(j, residue)
                    stack.append(k)
                    residue -= candidates[k]
            else:
                if residue == 0:
                    result.append([candidates[p] for p in reversed(stack)])
                while stack:
                    j = stack.pop()
                    residue += candidates[j]
                    if j:
                        stack.append(j - 1)
                        residue -= candidates[j - 1]
                        break
        return result

    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum2(self, candidates, target):
        candidates = sorted(candidates)

        def binarysearch(end, num):
            start = 0
            while start < end:
                mid = (start + end + 1) >> 1
                if candidates[mid] > num:
                    end = mid - 1
                else:
                    start = mid
            return start

        result, i = [], binarysearch(len(candidates) - 1, target)
        stack, residue = [i], target - candidates[i]
        while stack:
            i = stack[-1]
            if residue > 0 and i:
                k = min(binarysearch(i, residue), i - 1)
                stack.append(k)
                residue -= candidates[k]
            else:
                if residue == 0:
                    result.append([candidates[p] for p in reversed(stack)])
                while stack:
                    j = stack.pop()
                    residue, k = residue + candidates[j], j - 1
                    while k >= 0 and candidates[k] == candidates[k + 1]:
                        k -= 1
                    if k > -1:
                        stack.append(k)
                        residue -= candidates[k]
                        break
        return result


if __name__ == '__main__':
    comb = Solution()
    print(comb.combinationSum([1], 2))
    print(comb.combinationSum([1], 1))
    print(comb.combinationSum2([4, 4, 2, 1, 4, 2, 2, 1, 3], 6))

File: leetcode/求和/javascript/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
    scand = candidates.sort((a, b) => a - b);

    const bSearch = (end, pivot) => {
        let start = 0;
        while (start < end) {
            const mid = (start + end) >> 1;
            if (scand[mid] > pivot)
                end = mid;
            else
                start = mid + 1;
        }
        return start;
    }

    let result = [];
    let [stack, cur, residue] = [
        [], scand.length, target
    ];
    do {
        if (residue && cur) {
            cur = bSearch(cur, residue);
            if (cur) {
                cur--;
                const n = Math.floor(residue / scand[cur]);
                let temp = new Array(n);
                temp.fill(cur);
                stack.push(...temp);
                residue -= n * scand[cur];
            }
        } else {
            if (!residue) {
                let temp = stack.map(i => scand[i]);
                result.push(temp.reverse());
            }
            cur = stack.pop();
            residue += scand[cur];
        }
    } while (stack.length || cur);

    return result;
};

//recursion can also be used
var combinationSum1 = function (candidates, target) {
    let stack = [];
    let result = [];
    let [residue, cur] = [target, 0];
    while (stack.length || cur < candidates.length) {
        if (cur === candidates.length) {
            if (stack.length) {
                cur = stack.pop();
                residue += candidates[cur];
                cur++;
            }
        } else if (!residue) {
            result.push(stack.map(i => candidates[i]));
            cur = candidates.length;
        } else if (residue >= candidates[cur]) {
            stack.push(cur);
            residue -= candidates[cur];
        } else {
            cur++;
        }
    }
    return result;
};

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
    let scand = candidates.filter(i => i <= target).sort((a, b) => a - b);
    let result = [];
    let stack = [];
    let residue = target;
    let i = 0;
    while (stack.length || i < scand.length) {
        if (i === scand.length) {
            if (stack.length) {
                i = stack.pop();
                residue += scand[i];
                i++;
                while (scand[i] === scand[i - 1]) i++;
            }
        } else if (residue >= scand[i]) {
            stack.push(i);
            residue -= scand[i];
            if (!residue) {
                result.push(stack.map(j => scand[j]));
                i = scand.length;
            } else
                i++;
        } else {
            i = scand.length;
        }
    }
    return result;
};

var arr = [2, 3, 6, 7];
//console.log(combinationSum(arr, 7));
//console.log(combinationSum1(arr, 7));
var arr2 = [10, 1, 2, 7, 6, 1, 5];
console.log(combinationSum2(arr2, 8));